crates/solver/src/liquidity/zeroex.rs

```diff
@@ -77,6 +92,50 @@ impl ZeroExLiquidity {
         };
         Some(Liquidity::LimitOrder(limit_order))
     }
+
+    async fn run_orderbook_fetching(
+        api: Arc<dyn ZeroExApi>,
+        blocks_stream: CurrentBlockStream,
+        orderbook_cache: Arc<OrderbookCache>,
+        gpv2_address: H160,
+    ) {
+        let mut receiver = blocks_stream.clone();
+        loop {
+            let queries = &[
+                // orders fillable by anyone
+                OrdersQuery::default(),
+                // orders fillable only by our settlement contract
+                OrdersQuery {
+                    sender: Some(gpv2_address),
+                    ..Default::default()
+                },
+            ];
+            let zeroex_orders_results =
+                futures::future::join_all(queries.iter().map(|query| api.get_orders(query))).await;
+            let zeroex_orders = zeroex_orders_results
+                .into_iter()
+                .flat_map(|result| {
+                    result.unwrap_or_else(|err| {
+                        tracing::error!(?err, "ZeroExResponse error during liqudity fetching");
+                        vec![]
+                    })
+                })
+                .collect();
+
+            {
+                let mut cache = orderbook_cache.write().await;
+                *cache = zeroex_orders;
+            }
+
+            while let Err(err) = receiver.changed().await {
+                tracing::error!(
+                    ?err,
+                    "failed to receive block update during 0x liquidity fetching"
+                );
+                tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+            }
```

Comment:
Added this logic at the end of the function to allow the cache fetching on startup, without waiting for a new block.

Reply:
I don't think you need this if you turn the CurrentBlockStream into a stream and simply `await next` (cf. suggestion above). This is how I used it in :https://github.com/cowprotocol/services/blob/6b32c72e3290a6e76004ad5d6eef90b7bad24f90/crates/driver/src/domain/mempools.rs#L100-L104 (there I wanted to explicitly wait for a new block)